imports/wasi_snapshot_preview1/wrapper.go

```diff
@@ -0,0 +1,124 @@
+package wasi_snapshot_preview1
+
+import (
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/internal/leb128"
+	"github.com/tetratelabs/wazero/internal/wasm"
+)
+
+// mustProxyV_I32I32 is a convenience that panics on problems compiling
+// proxyV_I32I32.
+func mustProxyV_I32I32(goFunc *wasm.HostFunc, name string, paramNames ...string) *wasm.ProxyFunc {
+	if ret, err := proxyV_I32I32(goFunc, name, paramNames...); err != nil {
+		panic(err)
+	} else {
+		return ret
+	}
+}
+
+// memBytes adds an i32 value of memory[0]'s size in bytes to the stack
+var memBytes = append(append([]byte{wasm.OpcodeMemorySize, 0, wasm.OpcodeI32Const}, leb128.EncodeInt32(int32(wasm.MemoryPageSize))...), wasm.OpcodeI32Mul)
+
+const (
+	nilType     = 0x40
+	success     = byte(ErrnoSuccess)
+	fault       = byte(ErrnoFault)
+	fakeFuncIdx = byte(0)
+)
+
+// proxyV_I32I32 proxies the goFunc which returns takes no parameters and
+// returns two i32 results and an Errno (three total).
+//
+// The proxied function takes two i32 parameters, which are memory offsets to
+// write first two results on ErrnoSuccess. The Errno returned by goFunc is
+// returned from the proxied function, regardless.
+func proxyV_I32I32(proxied *wasm.HostFunc, name string, paramNames ...string) (*wasm.ProxyFunc, error) {
+	proxy := &wasm.HostFunc{
+		ExportNames: []string{name},
+		Name:        name,
+		ParamTypes:  []api.ValueType{i32, i32},
+		ResultTypes: []api.ValueType{i32},
+		ParamNames:  paramNames,
+		ResultNames: []string{"errno"},
+		Code:        &wasm.Code{IsHostFunction: true, LocalTypes: []api.ValueType{i32}},
+	}
+
+	// Determine the index of a temporary local used for i32 manipulation. This
+	// will be directly after the parameters.
+	localI32Temp := byte(len(proxy.ParamTypes))
+
+	// Get the current memory size in bytes, to validate parameters.
+	body := memBytes // stack: [memSizeBytes]
+
+	// Ensure param[0] and param[1] are valid memory offsets for storing i32.
```

Comment:
on my laptop, doing bounds validation in wasm makes it a few nanoseconds slower than before. However, I think the visibility is worth it.

An aside: not all wasi implementations actually check memory bounds, but at least one besides wazero does (wasmtime iirc does)